### PR TITLE
expression: allow json functions to be expression index

### PR DESCRIPTION
**Pull request by xiongjiwei**
Signed-off-by: xiongjiwei <xiongjiwei1996@outlook.com>

### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/37740

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: ddl/db_integration_test.go [hunk 1]
```diff
@@ -2467,6 +2467,52 @@ func TestAddExpressionIndex(t *testing.T) {
...
+	tk.MustGetErrCode("create table t(a char(255), index idx((json_quote(a))));", errno.ErrTooLongKey)
+	tk.MustExec("create table t(a char(40));")
+	tk.MustExec("insert into t values ('[1, 2, 3]')")
+	tk.MustExec("alter table t add index idx3((json_quote(a)));")
+	tk.MustQuery("select * from t force index(idx3) where json_quote(a) = '\"[1, 2, 3]\"';").Check(testkit.Rows("[1, 2, 3]"))
+	tk.MustQuery("select * from t ignore index(idx3) where json_quote(a) = '\"[1, 2, 3]\"';").Check(testkit.Rows("[1, 2, 3]"))
+
+	tk.MustExec("drop table if exists t;")
+	tk.MustExec("create table t(a int, b json);")
+	tk.MustGetErrCode(`alter table t add index idx_wrong((json_array(b)));`, errno.ErrFunctionalIndexOnJSONOrGeometryFunction)
```
Comment (wjhuang2016) [inline ddl/db_integration_test.go:2504]: Please add some success tests for these functions.

**Review by YangKeao — APPROVED** [review]
LGTM

File: ddl/db_integration_test.go (hunk 1 above)
Comment (xiongjiwei) [reply ddl/db_integration_test.go:2504]: more tests are added

**Review by wjhuang2016 — APPROVED** [review]

**Review by CbcWestwolf — APPROVED** [review]

**Review by bb7133 — APPROVED** [review]
LGTM

**Pull request merged**